packages/backend-common/src/tokens/AzureAccessToken.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { ClientSecretCredential } from '@azure/identity';
+import { TokenCredential } from '@azure/core-auth';
+import { InputError } from '@backstage/errors';
+
+/**
+ * Helpers singtone class to distunguiesh connection.
+ * @TODO remove before merge.
+ */
+export class ConnectionCounter {
+  private static _instance: ConnectionCounter;
+  public static counter: number = 0;
+  public static get Instance() {
+    return this._instance || (this._instance = new this());
+  }
+  public static increment(): number {
+    ConnectionCounter.counter = ConnectionCounter.counter + 1;
+    return ConnectionCounter.counter;
+  }
+}
+
+/**
+ * Retrieves Azure AD access token information
+ *
+ * @public
+ */
+export class AzureAccessToken {
+  public credential: TokenCredential;
+  public counter: number;
+  constructor(
+    private tenantId: string,
+    private clientId: string,
+    private clientSecret: string,
+  ) {
+    this.credential = new ClientSecretCredential(
+      this.tenantId,
+      this.clientId,
+      this.clientSecret,
+    );
+    this.counter = ConnectionCounter.increment(); // @TODO remove before merge.
+  }
+
+  /**
+   * Fetches access token information from Azure AD for RDBSM
+   */
+  public async getTokenInfo() {
+    try {
+      const tokenResponse = await this.credential.getToken(
+        'https://ossrdbms-aad.database.windows.net/.default',
+      );
+      return {
+        token: tokenResponse?.token || '',
+        expiresOnTimestamp: tokenResponse?.expiresOnTimestamp || '',
+        connectionNumber: this.counter, // @TODO remove before merge.
+      };
+    } catch (error) {
+      console.error('Error retrieving access token:', error.message);
+      throw error;
+    }
+  }
+
+  /**
+   * Validates Azure AD Auth credentials.
+   *
+   * @param azureCred
+   * @returns boolean
+   */
+  public static validateCredentials(azureCred: any) {
+    if (
+      typeof azureCred !== 'object' ||
+      'clientId' in azureCred === false ||
+      'tenantId' in azureCred === false ||
+      'clientSecret' in azureCred === false
+    ) {
+      throw new InputError(`Wrong Azure AD authentication credentials`);
+    }
+    return true;
+  }
```

Comment:
This is very similar to the [isAzureDevOpsCredential](https://github.com/XpiritBV/backstage/blob/d54778f84912e5ac100b3d62af41556d78c95fd2/packages/integration/src/azure/config.ts#L182) type predicate I added in #18213.

Since we are also dealing with Azure credentials here, it would be nice to unify both the reading from the config and the type narrowing so that the shape of the Azure credential configuration is the same for database connections and the Azure integration.

@awanlin, any thoughts on this?

Reply:
Thank you for the reference :)
Since [isAzureDevOpsCredential](https://github.com/XpiritBV/backstage/blob/d54778f84912e5ac100b3d62af41556d78c95fd2/packages/integration/src/azure/config.ts#L182) is a new feature that is waiting to be merged, and [isAzureClientSecretCredential ](https://github.com/backstage/backstage/blob/master/packages/integration/src/azure/config.ts#L88C14-L88C44) is not exported from `integration` package, I reimplemented `isAzureClientSecretCredential`  function in my PR. So there is at least the same approach used and it can be easily adjusted.

Can you help me find a way to avoid this duplication?

Reply:
I refactored this into a discriminated union based on feedback from @Rugvip. We'd have to move the config types, credential types and parser code to a common place so you can reuse it. I don't know if `integrations` is the correct package or whether a new package, like `azure-common`, `credentials`, or `authentication`, is required.

@Rugvip, @benjdlambert do you have any thoughts on this? Some additional context: any integration with Azure, like databases, Azure DevOps, plugins, can use the `@azure/identity` library to get an access token. So some shared config types and a way to map from config to a `TokenCredential` (interface for all Azure credentials) would be helpful I think

Reply:
I refactored fetching Azure token from `integration` package. I made some functionality I've used as public. However, the naming is a bit odd. DB authentication has nothing to do with DevOps but it's using functionality named like this.